sycl/doc/extensions/proposed/sycl_ext_oneapi_fpga_mem.asciidoc

```diff
@@ -0,0 +1,507 @@
+====
+
+The purpose of this document is to clearly describe and specify `fpga_mem` and 
+related concepts, types, and mechanisms, and to give examples and context for their usage.
+
+=== Motivation
+`fpga_mem` is a wrapper around user variables that asserts the variable should be implemented in FPGA memory resources and enables attaching compile-time hints which configure the memory resource and may impact the performance of the design.
+
+=== Example
+[source,c++]
+----
+struct MyClass {
+  bool x;
+};
+
+using namespace sycl::ext::intel::experimental;
+
+sycl::queue q;
+q.single_task([=] {
+  fpga_mem<int[4], decltype(properties(resource_mlab))> fm1{1, 3, 5, 7};
+  fpga_mem<MyClass, decltype(properties(resource_mlab))> fm2;
+  fm2.get().x = fm1[0];
+});
+----
+
+Above we see an example that declares `fpga_mem` variables `fm1`, `fm2`, and accesses them. 
+`fm1` creates an allocation on the device that is private to the kernel and contains an array of four `int`s that is initialized to {1, 3, 5, 7}. `fm2` creates an allocation on the device that is private to the kernel and contains an object of type `MyClass` that is default-initialized. Both variables will be implemented using the MLAB hardware resource. For more details on supported properties see a section below.
+
+`fpga_mem` has reference wrapper-like semantics, and is implicitly convertible to the wrapped type. Of note,
+because {cpp} doesn't allow for overloading of the "dot operator", a `get()`
+member of `fpga_mem` allows a reference to be extracted, to which the usual
+dot operator may be applied.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification. An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_FPGA_MEM` to one of the values defined in the table
+below. Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Representation of FPGA memory
+
+[source,c++]
+----
+namespace sycl::ext::intel::experimental {
+template <typename T, typename PropListT = sycl::ext::oneapi::properties<>>
+class fpga_mem {
+  ...
+----
+
+`fpga_mem` is a class template, parameterized by the type of the underlying
+allocation _T_, and a list of properties _PropListT_. The type of the
+allocation _T_ also encodes the size of the allocation.
+
+`fpga_mem` initializes the underlying object `T` with the values passed into its constructor.
+
+Properties may be specified for a `fpga_mem` to provide semantic
+modification or optimization hint information to the compiler. See the section
```

Comment:
Is "semantic modification" really the right term here when the behaviour of the program remains unchanged?